core/src/main/java/com/microsoft/applicationinsights/internal/jmx/JmxDataFetcher.java

```diff
@@ -83,45 +89,85 @@
+            Object obj = server.getAttribute(object, attributeName);
+            if (obj != null) {
+                result.add(obj);
+            } else {
+                InternalLogger.INSTANCE.warn("Could not find JMX attribute named '%s' in object '%s'", attributeName, object);
+            }
+        }
+    }
 
-            Object obj;
+    static void fetchCompositeObjects(MBeanServer server, Set<ObjectName> objects, String attributeName, List<Object> result)
+            throws AttributeNotFoundException, MBeanException, ReflectionException, InstanceNotFoundException {
 
-            if (innerAttributeType != AttributeType.REGULAR) {
-                obj = server.getAttribute(object, inners[0]);
+        String[] inners = attributeName.split("\\.");
+        for (ObjectName object : objects) {
+            CompositeDataSupport compositeData = (CompositeDataSupport) server.getAttribute(object, inners[0]);
+            if (compositeData == null) {
+                InternalLogger.INSTANCE.warn("Could not find composite attribute named '%s' for '%s' in object '%s'", inners[0], attributeName, object);
+                continue;
+            }
+            Object obj = compositeData.get(inners[1]);
 
-                javax.management.openmbean.CompositeDataSupport compositeData = null;
-                if (innerAttributeType == AttributeType.TABULAR) {
-                    javax.management.openmbean.TabularDataSupport tabularData = (javax.management.openmbean.TabularDataSupport)obj;
-                    compositeData = (CompositeDataSupport) tabularData.get(inners[1]);
-                    obj = compositeData.get(inners[2]);
-                } else {
-                    compositeData = (javax.management.openmbean.CompositeDataSupport)obj;
-                    obj = compositeData.get(inners[1]);
-                }
+            if (obj != null) {
+                result.add(obj);
             } else {
-                obj = server.getAttribute(object, attr);
+                InternalLogger.INSTANCE.warn("Could not find composite attribute named '%s' for '%s' in object '%s'", inners[1], attributeName, object);
+            }
+        }
+    }
+
+    static void fetchTabularObjects(MBeanServer server, Set<ObjectName> objects, String attributeName, List<Object> result)
+            throws AttributeNotFoundException, MBeanException, ReflectionException, InstanceNotFoundException {
+
+        String[] inners = attributeName.split("\\.");
+        for (ObjectName object : objects) {
+            TabularDataSupport tabularData = (TabularDataSupport) server.getAttribute(object, inners[0]);
+            if (tabularData == null) {
+                InternalLogger.INSTANCE.warn("Could not find tabular attribute named '%s' for '%s' in object '%s'", inners[0], attributeName, object);
+                continue;
             }
+
+            CompositeDataSupport compositeData = (CompositeDataSupport) tabularData.get(inners[1]);
```

Comment:
There's a bug here: [`TabularDataSupport.get(Object)`](https://docs.oracle.com/javase/7/docs/api/javax/management/openmbean/TabularDataSupport.html#get(java.lang.Object)) calls [`TabularDataSupport.get(Object[])`](https://docs.oracle.com/javase/7/docs/api/javax/management/openmbean/TabularDataSupport.html#get(java.lang.Object[])) just by casting `Object` to `Object[]`. I don't think we've seen issues here because it appears `TabularData` in MXBeans is pretty rare. I wasn't able to find an example and only followed the documentation.

This needs to be fixed before merging. Since `get` takes and array, this also probably means the current way to reference tabular data is insufficient. I'm going to fix the cast issue and try to find a bean in JConsole with TabularData to test against.